devices/flash-zynq7000/qspi.c

```diff
@@ -0,0 +1,329 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system loader
+ *
+ * Quad-SPI Controller driver
+ *
+ * Copyright 2020 Phoenix Systems
+ * Author: Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "qspi.h"
+#include <lib/errno.h>
+
+
+enum { cr = 0, sr, ier, idr, imr, er, dr, txd00, rxd, sicr, txth, rxth, gpio,
+	   lpbk = 0xe, txd01 = 0x20, txd10, txd11,
+	   lqspi_cr = 0x28, lqspi_sr, modid = 0x3f };
+
+
+struct {
+	volatile u32 *base;
+} qspi_common;
+
+
+void qspi_stop(void)
+{
+	u32 data;
+
+	/* Clean up RX Fifo */
+	while ((*(qspi_common.base + sr) & (1 << 4)))
+		data = *(qspi_common.base + rxd);
+
+	*(qspi_common.base + cr) |= (1 << 10);
+	*(qspi_common.base + er) = 0;
+}
+
+
+void qspi_start(void)
+{
+	*(qspi_common.base + rxth) = 0x1;
+	*(qspi_common.base + cr) &= ~(1 << 10);
+	*(qspi_common.base + er) = 0x1;
+}
+
+
+static unsigned int qspi_rxData(u8 *rxBuff, size_t size)
+{
+	const u32 data = *(qspi_common.base + rxd);
+
+	if (rxBuff != NULL) {
+		switch (size) {
+			case 0:
+				return 0;
+
+			case 1:
+				rxBuff[0] = (data >> 24) & 0xff;
+				return 1;
+
+			case 2:
+				rxBuff[0] = (data >> 16) & 0xff;
+				rxBuff[1] = (data >> 24) & 0xff;
+				return 2;
+
+			case 3:
+				rxBuff[0] = (data >> 8) & 0xff;
+				rxBuff[1] = (data >> 16) & 0xff;
+				rxBuff[2] = (data >> 24) & 0xff;
+				return 3;
+
+			case 4:
+			default:
+				rxBuff[0] = data & 0xff;
+				rxBuff[1] = (data >> 8) & 0xff;
+				rxBuff[2] = (data >> 16) & 0xff;
+				rxBuff[3] = (data >> 24) & 0xff;
+				return 4;
+		}
+	}
+
+	return (size < 4) ? size : 4;
```

Comment:
Dead code, also it will return size, when `rxBuff` was NULL (in what case it should return error or zero). Change returns to breaks in switch-case and add case for rxBuff == NULL